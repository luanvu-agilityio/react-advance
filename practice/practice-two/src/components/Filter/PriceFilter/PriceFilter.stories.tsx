import type { Meta, StoryObj } from '@storybook/react'
import { useState, type ChangeEvent } from 'react'
import { PriceFilter } from './PriceFilter'
import type { PriceRangeProps } from 'types/Filter'

const meta: Meta<typeof PriceFilter> = {
  title: 'Components/Filter/PriceFilter',
  component: PriceFilter,
  parameters: {
    layout: 'centered',
    backgrounds: {
      default: 'light',
    },
  },
  tags: ['autodocs'],
}

export default meta
type Story = StoryObj<typeof PriceFilter>

/**
 * Default PriceFilter with pre-set values
 */
export const Default: Story = {
  args: {
    priceRange: { min: 100, max: 500 },
    sliderValues: [100, 500],
    onSliderChange: (values) => {
      console.log(`Slider values changed: ${values}`)
    },
    onMinInputChange: (e) => {
      console.log(`Min input changed: ${e.target.value}`)
    },
    onMaxInputChange: (e) => {
      console.log(`Max input changed: ${e.target.value}`)
    },
    onApplyFilters: () => {
      console.log('Apply filters clicked')
    },
    onResetFilters: () => {
      console.log('Reset filters clicked')
    },
  },
}

/**
 * PriceFilter with minimum and maximum values set
 */
export const MinMaxValues: Story = {
  args: {
    priceRange: { min: 0, max: 1000 },
    sliderValues: [0, 1000],
    onSliderChange: () => {},
    onMinInputChange: () => {},
    onMaxInputChange: () => {},
    onApplyFilters: () => {},
    onResetFilters: () => {},
  },
}

/**
 * PriceFilter with narrow price range
 */
export const NarrowRange: Story = {
  args: {
    priceRange: { min: 450, max: 550 },
    sliderValues: [450, 550],
    onSliderChange: () => {},
    onMinInputChange: () => {},
    onMaxInputChange: () => {},
    onApplyFilters: () => {},
    onResetFilters: () => {},
  },
}

/**
 * Fully interactive PriceFilter that demonstrates state changes
 */
export const Interactive: Story = {
  render: function InteractivePriceFilter() {
    const [priceRange, setPriceRange] = useState<PriceRangeProps>({
      min: 200,
      max: 800,
    })
    const [sliderValues, setSliderValues] = useState([200, 800])
    const [appliedValues, setAppliedValues] = useState({
      min: 200,
      max: 800,
    })

    const handleSliderChange = (values: number[]) => {
      setSliderValues(values)
      setPriceRange({
        min: values[0],
        max: values[1],
      })
    }

    const handleMinInputChange = (e: ChangeEvent<HTMLInputElement>) => {
      const value = Math.min(Number(e.target.value), priceRange.max)
      setPriceRange({
        ...priceRange,
        min: value,
      })
      setSliderValues([value, sliderValues[1]])
    }

    const handleMaxInputChange = (e: ChangeEvent<HTMLInputElement>) => {
      const value = Math.max(Number(e.target.value), priceRange.min)
      setPriceRange({
        ...priceRange,
        max: value,
      })
      setSliderValues([sliderValues[0], value])
    }

    const handleApplyFilters = () => {
      setAppliedValues({
        min: sliderValues[0],
        max: sliderValues[1],
      })
      alert(`Applied filter: $${sliderValues[0]} - $${sliderValues[1]}`)
    }

    const handleResetFilters = () => {
      setPriceRange({ min: 0, max: 1000 })
      setSliderValues([0, 1000])
      setAppliedValues({ min: 0, max: 1000 })
      alert('Filters reset to default values')
    }

    return (
      <div>
        <PriceFilter
          priceRange={priceRange}
          sliderValues={sliderValues}
          onSliderChange={handleSliderChange}
          onMinInputChange={handleMinInputChange}
          onMaxInputChange={handleMaxInputChange}
          onApplyFilters={handleApplyFilters}
          onResetFilters={handleResetFilters}
        />
        <div style={{ marginTop: 20, padding: 10, backgroundColor: '#f5f5f5' }}>
          <h4>Current State:</h4>
          <p>
            Price Range: ${priceRange.min} - ${priceRange.max}
          </p>
          <p>
            Slider Values: ${sliderValues[0]} - ${sliderValues[1]}
          </p>
          <p>
            Applied Values: ${appliedValues.min} - ${appliedValues.max}
          </p>
        </div>
      </div>
    )
  },
}
