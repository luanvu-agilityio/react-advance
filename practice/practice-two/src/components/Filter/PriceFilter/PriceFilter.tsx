import { type ChangeEvent } from 'react'
import {
  SectionTitle,
  PriceSection,
  SliderRoot,
  SliderTrack,
  SliderRange,
  SliderThumb,
  PriceInputsContainer,
  PriceInputWrapper,
  PriceInputLabel,
  PriceInput,
  Separator,
  ButtonsContainer,
} from '../FilterStyles'
import type { PriceRangeProps } from 'types/Filter'
import { Button } from '@radix-ui/themes'

interface PriceFilterProps {
  priceRange: PriceRangeProps
  sliderValues: number[]
  onSliderChange: (values: number[]) => void
  onMinInputChange: (e: ChangeEvent<HTMLInputElement>) => void
  onMaxInputChange: (e: ChangeEvent<HTMLInputElement>) => void
  onApplyFilters: () => void
  onResetFilters: () => void
}

export const PriceFilter = ({
  priceRange,
  sliderValues,
  onSliderChange,
  onMinInputChange,
  onMaxInputChange,
  onApplyFilters,
  onResetFilters,
}: PriceFilterProps) => {
  return (
    <PriceSection>
      <SectionTitle>Price</SectionTitle>
      <SliderRoot
        value={sliderValues}
        onValueChange={onSliderChange}
        min={0}
        max={1000}
        step={1}
      >
        <SliderTrack>
          <SliderRange />
        </SliderTrack>
        <SliderThumb />
        <SliderThumb />
      </SliderRoot>

      <PriceInputsContainer>
        <PriceInputWrapper>
          <PriceInputLabel>Min</PriceInputLabel>
          <PriceInput
            type="number"
            value={priceRange.min}
            onChange={onMinInputChange}
            placeholder="0"
            min="0"
          />
        </PriceInputWrapper>
        <Separator>-</Separator>
        <PriceInputWrapper>
          <PriceInputLabel>Max</PriceInputLabel>
          <PriceInput
            type="number"
            value={priceRange.max}
            onChange={onMaxInputChange}
            placeholder="1000"
            min="0"
          />
        </PriceInputWrapper>
      </PriceInputsContainer>

      <ButtonsContainer>
        <Button variant="solid" onClick={onApplyFilters} size={'2'}>
          Apply
        </Button>
        <Button
          variant="soft"
          onClick={onResetFilters}
          style={{ color: 'var(--black-shade-2' }}
        >
          Reset
        </Button>
      </ButtonsContainer>
    </PriceSection>
  )
}
