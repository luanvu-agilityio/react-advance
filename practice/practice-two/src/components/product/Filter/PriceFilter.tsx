import { type ChangeEvent } from 'react'
import {
  SectionTitle,
  PriceSection,
  SliderRoot,
  SliderTrack,
  SliderRange,
  SliderThumb,
  PriceInputsContainer,
  PriceInputWrapper,
  PriceInputLabel,
  PriceInput,
  Separator,
  ButtonsContainer,
  ApplyButton,
  ResetButton,
} from './FilterStyles'
import type { PriceRangeProps } from 'types/Filter'

interface PriceFilterProps {
  priceRange: PriceRangeProps
  sliderValues: number[]
  onSliderChange: (values: number[]) => void
  onMinInputChange: (e: ChangeEvent<HTMLInputElement>) => void
  onMaxInputChange: (e: ChangeEvent<HTMLInputElement>) => void
  onApplyFilters: () => void
  onResetFilters: () => void
}

export const PriceFilter = ({
  priceRange,
  sliderValues,
  onSliderChange,
  onMinInputChange,
  onMaxInputChange,
  onApplyFilters,
  onResetFilters,
}: PriceFilterProps) => {
  return (
    <PriceSection>
      <SectionTitle>Price</SectionTitle>
      <SliderRoot
        value={sliderValues}
        onValueChange={onSliderChange}
        min={0}
        max={1000}
        step={1}
      >
        <SliderTrack>
          <SliderRange />
        </SliderTrack>
        <SliderThumb />
        <SliderThumb />
      </SliderRoot>

      <PriceInputsContainer>
        <PriceInputWrapper>
          <PriceInputLabel>Min</PriceInputLabel>
          <PriceInput
            type="number"
            value={priceRange.min}
            onChange={onMinInputChange}
            placeholder="0"
            min="0"
          />
        </PriceInputWrapper>
        <Separator>-</Separator>
        <PriceInputWrapper>
          <PriceInputLabel>Max</PriceInputLabel>
          <PriceInput
            type="number"
            value={priceRange.max}
            onChange={onMaxInputChange}
            placeholder="1000"
            min="0"
          />
        </PriceInputWrapper>
      </PriceInputsContainer>

      <ButtonsContainer>
        <ApplyButton onClick={onApplyFilters}>Apply</ApplyButton>
        <ResetButton onClick={onResetFilters}>Reset</ResetButton>
      </ButtonsContainer>
    </PriceSection>
  )
}
