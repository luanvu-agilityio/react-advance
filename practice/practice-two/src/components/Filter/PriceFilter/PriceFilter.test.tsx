import { render, screen, fireEvent } from '@testing-library/react'
import { PriceFilter } from './PriceFilter'
import type { PriceRangeProps } from 'types/Filter'
import type { ReactNode } from 'react'

// Mock the slider component since it might use complex DOM interactions
jest.mock('../FilterStyles', () => {
  const actual = jest.requireActual('../FilterStyles')
  return {
    ...actual,
    SliderRoot: ({
      children,
      onValueChange,
      value,
    }: {
      children: ReactNode
      onValueChange: (value: number[]) => void
      value: number[]
    }) => (
      <div
        data-testid="slider-root"
        onClick={() => onValueChange([value[0] + 10, value[1] - 10])}
        data-min={value[0]}
        data-max={value[1]}
      >
        {children}
      </div>
    ),
    SliderTrack: ({ children }: { children: ReactNode }) => (
      <div data-testid="slider-track">{children}</div>
    ),
    SliderRange: () => <div data-testid="slider-range" />,
    SliderThumb: () => <div data-testid="slider-thumb" />,
    PriceInput: (props: { 'aria-label'?: string; [key: string]: unknown }) => (
      <input
        data-testid={`price-input-${props['aria-label']?.toLowerCase()}`}
        {...props}
      />
    ),
  }
})

describe('PriceFilter', () => {
  const defaultProps = {
    priceRange: { min: 100, max: 500 } as PriceRangeProps,
    sliderValues: [100, 500],
    onSliderChange: jest.fn(),
    onMinInputChange: jest.fn(),
    onMaxInputChange: jest.fn(),
    onApplyFilters: jest.fn(),
    onResetFilters: jest.fn(),
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  // SNAPSHOT TESTS FOR RENDERING
  describe('Rendering', () => {
    it('renders different states correctly', () => {
      // Default state
      const { container: defaultContainer } = render(
        <PriceFilter {...defaultProps} />
      )
      expect(defaultContainer).toMatchSnapshot('default')

      // With different price range values
      const { container: differentRangeContainer } = render(
        <PriceFilter
          {...defaultProps}
          priceRange={{ min: 50, max: 700 }}
          sliderValues={[50, 700]}
        />
      )
      expect(differentRangeContainer).toMatchSnapshot('different-range')

      // With high range values
      const { container: extremeValuesContainer } = render(
        <PriceFilter
          {...defaultProps}
          priceRange={{ min: 0, max: 1000 }}
          sliderValues={[0, 1000]}
        />
      )
      expect(extremeValuesContainer).toMatchSnapshot('extreme-values')

      // With equal min/max (edge case)
      const { container: equalValuesContainer } = render(
        <PriceFilter
          {...defaultProps}
          priceRange={{ min: 250, max: 250 }}
          sliderValues={[250, 250]}
        />
      )
      expect(equalValuesContainer).toMatchSnapshot('equal-values')
    })
  })

  // INTERACTIVE BEHAVIOR TESTS
  describe('Interactive Behavior', () => {
    it('calls onSliderChange when slider value changes', () => {
      render(<PriceFilter {...defaultProps} />)

      const slider = screen.getByTestId('slider-root')
      fireEvent.click(slider)

      expect(defaultProps.onSliderChange).toHaveBeenCalledTimes(1)
      expect(defaultProps.onSliderChange).toHaveBeenCalledWith([110, 490])
    })

    it('calls onApplyFilters when Apply button is clicked', () => {
      render(<PriceFilter {...defaultProps} />)

      const applyButton = screen.getByText('Apply')
      fireEvent.click(applyButton)

      expect(defaultProps.onApplyFilters).toHaveBeenCalledTimes(1)
    })

    it('calls onResetFilters when Reset button is clicked', () => {
      render(<PriceFilter {...defaultProps} />)

      const resetButton = screen.getByText('Reset')
      fireEvent.click(resetButton)

      expect(defaultProps.onResetFilters).toHaveBeenCalledTimes(1)
    })
  })
})
