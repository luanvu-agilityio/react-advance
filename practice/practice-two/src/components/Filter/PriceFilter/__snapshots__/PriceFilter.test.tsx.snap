// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PriceFilter Rendering renders different states correctly: default 1`] = `
<div>
  <div
    class="sc-etfXYe lcTFgo"
  >
    <h2
      class="sc-fYsHOw gpnbOD"
    >
      Price
    </h2>
    <div
      data-max="500"
      data-min="100"
      data-testid="slider-root"
    >
      <div
        data-testid="slider-track"
      >
        <div
          data-testid="slider-range"
        />
      </div>
      <div
        data-testid="slider-thumb"
      />
      <div
        data-testid="slider-thumb"
      />
    </div>
    <div
      class="sc-dKKIkQ bENJTs"
    >
      <div
        class="sc-fpEFIB dKvHtD"
      >
        <label
          class="sc-gAqISa gDcqMA"
        >
          Min
        </label>
        <input
          data-testid="price-input-undefined"
          min="0"
          placeholder="0"
          type="number"
          value="100"
        />
      </div>
      <span
        class="sc-fbguzk bkCnXF"
      >
        -
      </span>
      <div
        class="sc-fpEFIB dKvHtD"
      >
        <label
          class="sc-gAqISa gDcqMA"
        >
          Max
        </label>
        <input
          data-testid="price-input-undefined"
          min="0"
          placeholder="1000"
          type="number"
          value="500"
        />
      </div>
    </div>
    <div
      class="sc-gGarWV bpubAX"
    >
      <button
        class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button"
        data-accent-color=""
      >
        Apply
      </button>
      <button
        class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-soft rt-Button"
        data-accent-color=""
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;

exports[`PriceFilter Rendering renders different states correctly: different-range 1`] = `
<div>
  <div
    class="sc-etfXYe lcTFgo"
  >
    <h2
      class="sc-fYsHOw gpnbOD"
    >
      Price
    </h2>
    <div
      data-max="700"
      data-min="50"
      data-testid="slider-root"
    >
      <div
        data-testid="slider-track"
      >
        <div
          data-testid="slider-range"
        />
      </div>
      <div
        data-testid="slider-thumb"
      />
      <div
        data-testid="slider-thumb"
      />
    </div>
    <div
      class="sc-dKKIkQ bENJTs"
    >
      <div
        class="sc-fpEFIB dKvHtD"
      >
        <label
          class="sc-gAqISa gDcqMA"
        >
          Min
        </label>
        <input
          data-testid="price-input-undefined"
          min="0"
          placeholder="0"
          type="number"
          value="50"
        />
      </div>
      <span
        class="sc-fbguzk bkCnXF"
      >
        -
      </span>
      <div
        class="sc-fpEFIB dKvHtD"
      >
        <label
          class="sc-gAqISa gDcqMA"
        >
          Max
        </label>
        <input
          data-testid="price-input-undefined"
          min="0"
          placeholder="1000"
          type="number"
          value="700"
        />
      </div>
    </div>
    <div
      class="sc-gGarWV bpubAX"
    >
      <button
        class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button"
        data-accent-color=""
      >
        Apply
      </button>
      <button
        class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-soft rt-Button"
        data-accent-color=""
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;

exports[`PriceFilter Rendering renders different states correctly: equal-values 1`] = `
<div>
  <div
    class="sc-etfXYe lcTFgo"
  >
    <h2
      class="sc-fYsHOw gpnbOD"
    >
      Price
    </h2>
    <div
      data-max="250"
      data-min="250"
      data-testid="slider-root"
    >
      <div
        data-testid="slider-track"
      >
        <div
          data-testid="slider-range"
        />
      </div>
      <div
        data-testid="slider-thumb"
      />
      <div
        data-testid="slider-thumb"
      />
    </div>
    <div
      class="sc-dKKIkQ bENJTs"
    >
      <div
        class="sc-fpEFIB dKvHtD"
      >
        <label
          class="sc-gAqISa gDcqMA"
        >
          Min
        </label>
        <input
          data-testid="price-input-undefined"
          min="0"
          placeholder="0"
          type="number"
          value="250"
        />
      </div>
      <span
        class="sc-fbguzk bkCnXF"
      >
        -
      </span>
      <div
        class="sc-fpEFIB dKvHtD"
      >
        <label
          class="sc-gAqISa gDcqMA"
        >
          Max
        </label>
        <input
          data-testid="price-input-undefined"
          min="0"
          placeholder="1000"
          type="number"
          value="250"
        />
      </div>
    </div>
    <div
      class="sc-gGarWV bpubAX"
    >
      <button
        class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button"
        data-accent-color=""
      >
        Apply
      </button>
      <button
        class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-soft rt-Button"
        data-accent-color=""
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;

exports[`PriceFilter Rendering renders different states correctly: extreme-values 1`] = `
<div>
  <div
    class="sc-etfXYe lcTFgo"
  >
    <h2
      class="sc-fYsHOw gpnbOD"
    >
      Price
    </h2>
    <div
      data-max="1000"
      data-min="0"
      data-testid="slider-root"
    >
      <div
        data-testid="slider-track"
      >
        <div
          data-testid="slider-range"
        />
      </div>
      <div
        data-testid="slider-thumb"
      />
      <div
        data-testid="slider-thumb"
      />
    </div>
    <div
      class="sc-dKKIkQ bENJTs"
    >
      <div
        class="sc-fpEFIB dKvHtD"
      >
        <label
          class="sc-gAqISa gDcqMA"
        >
          Min
        </label>
        <input
          data-testid="price-input-undefined"
          min="0"
          placeholder="0"
          type="number"
          value="0"
        />
      </div>
      <span
        class="sc-fbguzk bkCnXF"
      >
        -
      </span>
      <div
        class="sc-fpEFIB dKvHtD"
      >
        <label
          class="sc-gAqISa gDcqMA"
        >
          Max
        </label>
        <input
          data-testid="price-input-undefined"
          min="0"
          placeholder="1000"
          type="number"
          value="1000"
        />
      </div>
    </div>
    <div
      class="sc-gGarWV bpubAX"
    >
      <button
        class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button"
        data-accent-color=""
      >
        Apply
      </button>
      <button
        class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-soft rt-Button"
        data-accent-color=""
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;
